features/lorawan/LoRaWANStack.cpp

```diff
@@ -0,0 +1,2046 @@
+        _app_port = INVALID_PORT;
+    }
+
+#else
+    // initialize it to INVALID_PORT (255) an illegal port number.
+    // user should set the port
+     _app_port = INVALID_PORT;
+#endif
+
+     memset(&_lw_session, 0, sizeof(_lw_session));
+     memset(&_tx_msg, 0, sizeof(_tx_msg));
+     memset(&_rx_msg, 0, sizeof(_rx_msg));
+}
+
+LoRaWANStack::~LoRaWANStack()
+{
+
+}
+
+/*****************************************************************************
+ * Public member functions                                                   *
+ ****************************************************************************/
+LoRaWANStack& LoRaWANStack::get_lorawan_stack()
+{
+    static LoRaWANStack _lw_stack;
+    return _lw_stack;
+}
+
+radio_events_t *LoRaWANStack::bind_radio_driver(LoRaRadio& radio)
+{
+    // Store pointer to callback routines inside MAC layer (non-IRQ safe)
+    _mac_handlers = GetPhyEventHandlers();
+    //  passes the reference to radio driver down to PHY layer
+    lora_phy.set_radio_instance(radio);
+    return _mac_handlers;
+}
+lora_mac_status_t LoRaWANStack::initialize_mac_layer()
+{
+    if (DEVICE_STATE_NOT_INITIALIZED != _device_current_state)
+    {
+        tr_debug("Initialized already");
+        return LORA_MAC_STATUS_OK;
+    }
+
+    static LoRaMacPrimitives_t LoRaMacPrimitives;
+    static LoRaMacCallback_t LoRaMacCallbacks;
+    static lora_mac_mib_request_confirm_t mib_req;
+    static uint8_t compliance_test_buffer[LORAMAC_PHY_MAXPAYLOAD];
+
+    tr_debug("Initializing MAC layer");
+
+    // Allocate memory for compliance test
+    _compliance_test.app_data_buffer = compliance_test_buffer;
+
+    TimerTimeCounterInit( );
+    LoRaMacPrimitives.MacMcpsConfirm = callback(this, &LoRaWANStack::mcps_confirm);
+    LoRaMacPrimitives.MacMcpsIndication = callback(this, &LoRaWANStack::mcps_indication);
+    LoRaMacPrimitives.MacMlmeConfirm = callback(this, &LoRaWANStack::mlme_confirm);
+    LoRaMacCallbacks.TxNextPacketTimerEvent = callback(this, &LoRaWANStack::on_tx_next_packet_timer_event);
+    LoRaMacInitialization(&LoRaMacPrimitives, &LoRaMacCallbacks, lora_phy);
+
+    mib_req.type = LORA_MIB_ADR;
+    mib_req.param.adr_enable = LORAWAN_ADR_ON;
+    mib_set_request(&mib_req);
+
+    mib_req.type = LORA_MIB_PUBLIC_NETWORK;
+    mib_req.param.enable_public_network = LORAWAN_PUBLIC_NETWORK;
+    mib_set_request(&mib_req);
+
+    // Reset counters to zero. Will change in future with 1.1 support.
+    _lw_session.downlink_counter = 0;
+    _lw_session.uplink_counter = 0;
+
+    // Start loRaWAN state machine.
+    set_device_state(DEVICE_STATE_INIT);
+    return lora_state_machine();
+}
+
+/**
+ *
+ * Prepares the upload message to reserved ports
+ *
+ * \param port              Application port
+ * \return                  The length of data sent or a negative
+ *                          error code.
+ */
+int16_t LoRaWANStack::prepare_special_tx_frame(uint8_t port)
+{
+    switch (port) {
+    case 224:
```

Comment:
Why do we need switch/case construct to handle just one case?

Reply:
This part of the code base is incomplete as the Compliance Testing has not started yet. This method is part of Compliance Testing.